spices/SPICE-0009-external-readers.adoc

```diff
@@ -0,0 +1,120 @@
+= External Readers
+
+* Proposal: link:./SPICE-0009-external-readers.adoc[SPICE-0009]
+* Status: TBD
+* Implemented in: Pkl 0.27 ()
+* Category: Language
+
+== Introduction
+
+This SPICE proposes a way for custom module and resource readers to be used when evaluating modules with `pkl eval`.
+
+== Motivation
+
+When used via language binding libraries, custom module and resource readers schemes may be implemented by client code.
+Custom readers help bridge Pkl into systems where it does't fit natively, enabling a wide variety of use cases:
+
+* Mediating access to resources that require authentication (eg. a secret store)
+* Querying data accessible via non-HTTP protocols (eg. LDAP)
+* Providing filesystem abstractions for remote or in-memory modules
+* And more!
+
+One of the primary drawbacks to custom readers is that using one instantly turns a Pkl module into a _incompatible language_: the module may only be evaoluated by the client tool or application providing the reader and can no longer be evaluated directly using `pkl eval`.
+This hinders development and debugging of such modules, as typical workflows involving `trace(<expr>)` or `pkl eval -x '<expr>'` behave differently or are no longer possible.
+
+== Proposed Solution
+
+The existing message passing API and client libraries will be expanded to support external readers.
+Instead of client libraries invoking `pkl server`, `pkl eval` invocations, when configured with an external reader, will run the reader executable as a subprocess.
+External readers can provide any number of resource and/or module schemes and will configured as an executable and list of arguments via:
+
+* CLI flag
+* PklProject property
+* Java API
+* Message passing API
+
+
+[source,pkl]
+----
+/// May be specified as an URI to a readable resource, eg. `file:`, `https:`, or `package:`
+/// May also be specified as just an executable name, in which case it will be resolved according to the PATH environment variable
+executable: String
+
+/// Command line arguments that will be passed to the reader process
+arguments: Listing<String>
+----
+
+=== Example
+
+Consider this module:
+
+[source,pkl]
+----
+username = "john_appleseed"
+
+email = read("ldap://ds.example.com:389/dc=example,dc=com?mail?sub?(uid=\(username))").text
+----
+
+Pkl doesn't implement the `ldap:` resource URI scheme natively, but an external reader can provide it.
+Assuming a hypothetical `pkl-ldap` binary implementing the external reader protocol and the `ldap:` scheme is in the `$PATH`, this module can be evaluated as:
+
+[source,text]
+----
+$ pkl eval <module> --external-reader=pkl-ldap --allowed-resources prop:,ldap:
```

Comment:
In this example it is an executable, as briefly described under Proposed Solution:
```
/// May also be specified as just an executable name, in which case it will be resolved according to the PATH environment variable
executable: String
```

The end-to-end flow, as proposed, would be:
1. User invokes `pkl eval <module> --external-reader=pkl-ldap --allowed-resources prop:,ldap:`
2. `pkl` looks up `pkl-ldap` in `$PATH` and invokes it
3. `pkl` sends the `DiscoverReadersRequest` message
4. `pkl-ldap` responds with `DiscoverReadersResponse`. This may register zero or more resource or module reader schemes. In this case, it would make sense to register resource readers for the `ldap` and `ldaps` schemes.
5. Evaluation begins
6. Upon encountering `read("ldap:...")`, `pkl` sends a `ReadResourceRequest` (the same as it would to a language bindings client resource reader today)
7. The external reader process responds with `ReadResourceResponse`
8. Evaluation continues

Reply:
> pkl sends the DiscoverReadersRequest message

Ah, that was the missing piece.

This would mean for every external reader, Pkl would have to call it at least twice: at startup and then again for the actual read. Not a fan of the performance problems this might bring.
I think we can go with a simpler process of configuring the reader scheme statically. Something like (just an example, the syntax can be more ergonomic):

```bash
pkl eval <module> --external-reader ldap=pkl-ldap
```
Which wouldn't require Pkl to discover schemes at runtime.